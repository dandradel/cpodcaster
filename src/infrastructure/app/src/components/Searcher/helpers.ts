import { AnyAction, Dispatch } from '@reduxjs/toolkit'
import { Podcast } from '../../../../../domain/models/Podcast'
import { setPodcasts } from '../../store/reducers/podcastsSlice'

export interface FilterPodcastsArgs {
  stringSearch: string
  localPodcasts: Podcast[]
  dispatch: Dispatch<AnyAction>
}

export interface CheckStringInLabelArgs {
  label: string
  stringSearch: string
}

export const checkStringInLabel = ({ label, stringSearch }: CheckStringInLabelArgs): boolean =>
  label.toLowerCase().includes(stringSearch.toLowerCase())

export const filterPodcasts = ({
  stringSearch,
  localPodcasts,
  dispatch,
}: FilterPodcastsArgs): void => {
  if (stringSearch.length === 0) {
    dispatch(setPodcasts(localPodcasts))
    return
  }

  const filteredPodcasts: Podcast[] = []

  localPodcasts?.filter((podcast: Podcast) => {
    const { name, author } = podcast
    const podcastNameExists = checkStringInLabel({ label: author, stringSearch })
    const podcastAuthorExists = checkStringInLabel({
      label: name,
      stringSearch,
    })

    if (podcastNameExists || podcastAuthorExists) {
      filteredPodcasts?.push(podcast)
    }
  })

  dispatch(setPodcasts(filteredPodcasts))
}
