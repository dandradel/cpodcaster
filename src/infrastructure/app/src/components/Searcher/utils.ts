import { SxProps } from '@mui/material'

export const searcherBoxStyles: SxProps = {
  display: 'flex',
  flexDirection: 'row-reverse',
  p: 1,
  m: 1,
  bgcolor: 'background.paper',
  borderRadius: 1,
  gap: 1,
}
export const searcherTextFieldStyles: SxProps = {
  width: '20rem',
}
export const searcherChipStyles: SxProps = {
  my: 1,
  fontSize: '1.3rem',
  fontWeight: 'bold',
  borderRadius: 2,
}
