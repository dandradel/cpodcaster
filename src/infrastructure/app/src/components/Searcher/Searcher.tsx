/* eslint-disable @typescript-eslint/no-explicit-any */
import React, { ReactElement, useCallback } from 'react'
import { Chip, Box, TextField } from '@mui/material'
import { searcherBoxStyles, searcherChipStyles, searcherTextFieldStyles } from './styles'
import { useDispatch } from 'react-redux'
import useLocalStorage from '../../shared/hooks/useLocalStorage'
import { filterPodcasts } from './helpers'

interface SearcherProps {
  chipLabel: number | string | undefined
}

function Searcher({ chipLabel }: SearcherProps): ReactElement {
  const [localPodcasts] = useLocalStorage('podcasts', [])
  const dispatch = useDispatch()

  const handleSearcherChange = useCallback((e: any): void => {
    const stringSearch = e?.target.value

    filterPodcasts({ stringSearch, localPodcasts, dispatch })
  }, [localPodcasts, dispatch])

  return (
    <Box sx={searcherBoxStyles} data-testid='searcher-wrapper-testid'>
      <TextField
        label='Filter podcasts...'
        size='small'
        sx={searcherTextFieldStyles}
        onChange={handleSearcherChange}
      />
      <Chip
        color='primary'
        label={chipLabel}
        size='small'
        sx={searcherChipStyles}
        data-testid='search-chip-number-testid'
      />
    </Box>
  )
}

export default Searcher
