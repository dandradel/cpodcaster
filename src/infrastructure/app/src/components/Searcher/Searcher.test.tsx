/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react'
import '@testing-library/jest-dom'
import { cleanup, renderWithProviders, screen } from '../../shared/test/test-utils'
import userEvent from '@testing-library/user-event'
import Searcher from './Searcher'
import podcastsMock from '../../shared/__mocks__/podcasts.json'

describe('Searcher', () => {
  afterEach(cleanup)

  const setup = () => {
    const chipNumber = podcastsMock.length
    const utils = renderWithProviders(<Searcher chipLabel={chipNumber} />)
    const searcherInputEle = screen.getByLabelText('Filter podcasts...')
    const searcherChipEle = screen.getByText(chipNumber)
    return {
      searcherInputEle,
      searcherChipEle,
      chipNumber,
      userEvent,
      ...utils,
    }
  }

  test('renders Searcher', () => {
    const { container } = setup()
    expect(container).toMatchSnapshot()
  })
  test('renders Searcher with TextField and Chip components', () => {
    const { searcherInputEle, searcherChipEle } = setup()
    expect(searcherInputEle).toBeInTheDocument()
    expect(searcherChipEle).toBeInTheDocument()
  })
  test('renders Searcher and type in TextField', () => {
    const { searcherInputEle, userEvent } = setup()
    expect(searcherInputEle).toHaveValue('')
    userEvent.type(searcherInputEle, 'test')
    expect(searcherInputEle).toHaveValue('test')
  })
  test('renders Searcher and type in TextField and check Chip', () => {
    const { searcherInputEle, searcherChipEle, chipNumber, userEvent } = setup()
    expect(searcherInputEle).toHaveValue('')
    expect(searcherChipEle).toHaveTextContent(chipNumber as any)
    userEvent.type(searcherInputEle, 'test')
    expect(searcherInputEle).toHaveValue('test')
    expect(searcherChipEle).toHaveTextContent(chipNumber as any)
  })
})
