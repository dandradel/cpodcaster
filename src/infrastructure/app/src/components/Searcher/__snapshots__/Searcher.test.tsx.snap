// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Searcher renders Searcher 1`] = `
<div>
  <div
    class="MuiBox-root css-zyoky0"
    data-testid="searcher-wrapper-testid"
  >
    <div
      class="MuiFormControl-root MuiTextField-root css-x66ii1-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for=":r0:"
        id=":r0:-label"
      >
        Filter podcasts...
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r0:"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-yjsfm1"
          >
            <span>
              Filter podcasts...
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiChip-root MuiChip-filled MuiChip-sizeSmall MuiChip-colorPrimary MuiChip-filledPrimary css-vkz45j-MuiChip-root"
      data-testid="search-chip-number-testid"
    >
      <span
        class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
      >
        1
      </span>
    </div>
  </div>
</div>
`;
